Add StorageError and base exception handling

hasier opened the pull request:
We have encountered some errors with the Redis instances we use `limits` with, and `limits` bubbles those exceptions up. This is not bad on its own, but we are having to hardcode Redis exception handling in pieces of our code that may not be using Redis for their `limits` item storage, as we sometimes swap the type of storage for some of our endpoints. We thought it might be an interesting idea to introduce a `limits` exception to signal a storage exception to wrap the underlying one and keep a better separation of concerns this way.

A few notes:
- The idea is for each storage to set a base exception. At least one of the libraries does not have a hierarchy with just 1 at the top, so I used a tuple instead.
- I added the exception handling in the strategies, thinking that it might be useful that the storages raise the underlying exceptions. This might not be the desired approach, happy to move exception handling within the storage functions instead if that's preferred.
- I've relied on exceptions exposed in the main module for the dependencies, except for a couple of cases where I had to use `get_dependency()`. Please let me know if there's a different preferred approach.
- The idea for the future would be that we added more `limits`-specific errors inheriting from `StorageError`, and then each storage could map their own internal exceptions to those by overriding `get_storage_error()`.

I can add tests once it's confirmed this is a desirable behaviour and approach 🙂

hasier commented:
@alisaifee any chance you can look at this one? I appreciate this might not be a very high priority, but it feels like a useful abstraction to have considering the backends are easy to swap. And I could benefit myself of having something along these lines 🙂 

I know currently coverage checks are not passing, I just wanted to make sure this is a desirable approach before investing time in properly adding tests. Happy to hear your thoughts!

alisaifee commented:
> @alisaifee any chance you can look at this one? I appreciate this might not be a very high priority, but it feels like a useful abstraction to have considering the backends are easy to swap. And I could benefit myself of having something along these lines 🙂
> 
> I know currently coverage checks are not passing, I just wanted to make sure this is a desirable approach before investing time in properly adding tests. Happy to hear your thoughts!

My apologies for the delay in getting back to you. I'll review this in some detail today and try to provide some feedback - but conceptually this looks like a good addition to the library!

alisaifee commented:
> I added the exception handling in the strategies, thinking that it might be useful that the storages raise the underlying exceptions. This might not be the desired approach, happy to move exception handling within the storage functions instead if that's preferred.

Ideally I would want to keep this encapsulated within the `storage` namespace - though I can see how it might be a bit more cumbersome given the spread of the implementations for the storages. As you can see applying this in the strategies requires some duplication as well.

The first idea that comes to my mind is to perhaps use a decorator that wraps around the public methods of the storage implementations (though that too would incur some amount of duplication). What do you think?

alisaifee commented:
One minor issue here is that this change will be somewhat backward incompatible (for example if someone was explicitly catching exceptions from a specific storage class - they would no longer be caught). Perhaps pushing this handling all the way into storage might also provide a path to make this backward compatible by adding a `wrap_exceptions` option to all storage classes which control whether the original exception will be bubbled up as is or wrapped.

alisaifee reviewed (commented):
Left some thoughts/ideas in the comments for your consideration.

hasier commented:
Thanks for the comments @alisaifee!

> > I added the exception handling in the strategies, thinking that it might be useful that the storages raise the underlying exceptions. This might not be the desired approach, happy to move exception handling within the storage functions instead if that's preferred.
> 
> Ideally I would want to keep this encapsulated within the `storage` namespace - though I can see how it might be a bit more cumbersome given the spread of the implementations for the storages. As you can see applying this in the strategies requires some duplication as well.
> 
> The first idea that comes to my mind is to perhaps use a decorator that wraps around the public methods of the storage implementations (though that too would incur some amount of duplication). What do you think?

Yeah that's fair. Adding a decorator felt like duplicating a bit too much indeed. I tried to do a more blanket update by wrapping the functions within `__new__`, not sure if this overcomplicates things though, would love to know what you think. Added some basic tests as well to show how it works.

> One minor issue here is that this change will be somewhat backward incompatible (for example if someone was explicitly catching exceptions from a specific storage class - they would no longer be caught). Perhaps pushing this handling all the way into storage might also provide a path to make this backward compatible by adding a `wrap_exceptions` option to all storage classes which control whether the original exception will be bubbled up as is or wrapped.

Ah indeed! I have added that as well. I had to explicitly pass it up the `super()` calls to avoid `mypy` complaining, hopefully that's ok.

----

I also added some `Any` types and had to ignore them as `mypy` was complaining. Let me know if you'd rather handle them differently.

----

Besides, not sure what's going on with CI. The branch seems to be up to date with `master`, but the linter locally succeeds and fails in CI 🤔 Also, there still seems to be an issue with `pytest-asyncio` both here and in `master`.

alisaifee commented:
> Besides, not sure what's going on with CI. The branch seems to be up to date with master, but the linter locally succeeds and fails in CI 🤔 Also, there still seems to be an issue with pytest-asyncio both here and in master.

Sorry about that - I hadn't been paying attention for a bit and the state of pytest / pytest-asyncio is a bit in flux. I've fixed it in master (again) so please update against that.

Regarding linting - it's also because black has seen some updates so you might need to do:
`pip install -U -r requirements/dev.txt` to see the errors locally.

hasier commented:
> > Besides, not sure what's going on with CI. The branch seems to be up to date with master, but the linter locally succeeds and fails in CI 🤔 Also, there still seems to be an issue with pytest-asyncio both here and in master.
> 
> Sorry about that - I hadn't been paying attention for a bit and the state of pytest / pytest-asyncio is a bit in flux. I've fixed it in master (again) so please update against that.
> 
> Regarding linting - it's also because black has seen some updates so you might need to do: `pip install -U -r requirements/dev.txt` to see the errors locally.

No worries! I've also been having trouble with `pytest-asyncio` lately, so I completely understand. Thanks for fixing it!

I have just merged `master`, updated deps and run linter fixes, now everything seems to be passing 🎉 With that out of the way, I would love to hear your thoughts on [the new approach](https://github.com/alisaifee/limits/pull/192#issuecomment-1927040194) whenever you have a bit of time 🙂

alisaifee reviewed (approved):
The design looks like it checks all the boxes in my opinion. 
There are some aesthetic changes I might make and add more documentation & tests, but I can take care of that after merging to master.

alisaifee commented:
@hasier I'm going to merge this to master and perhaps tag you in a PR if I decide to make any changes from the public API perspective before making a release.

alisaifee commented:
@hasier I didn't end up with much changes to warrant a separate PR so have pushed straight to master. Mostly some reductions in code, typing & documentation tweaks. You can also take a look at the final documentation [here](https://limits.readthedocs.org/en/latest)

hasier commented:
> @hasier I didn't end up with much changes to warrant a separate PR so have pushed straight to master. Mostly some reductions in code, typing & documentation tweaks. You can also take a look at the final documentation [here](https://limits.readthedocs.org/en/latest)

Thanks for that @alisaifee, changes look good! Regarding docs, I was thinking of adding a new paragraph or maybe some mention to the new exception in the [Storage Backends](https://limits.readthedocs.io/en/latest/storage.html) section to make this behaviour more visible. But it might be a bit too much, as I said it's looking good, thank you! 🙇

alisaifee commented:
This is now available in [3.8.0](https://github.com/alisaifee/limits/releases/tag/3.8.0)
